modules/olvis3/data/templates/filter/src/ThresholdFilter.cpp

```diff
@@ -39,21 +41,35 @@ ThresholdFilter::ThresholdFilter()
     addInputPort(mThreshold);
 
     mMode.setName("mode");
-    mMode.setDesc("Thresholding mode");
-    mMode.addChoice(cv::THRESH_BINARY, "Binary (value = value > threshold ? max_value : 0 )");
-    mMode.addChoice(cv::THRESH_BINARY_INV, "Binary inverted (value = value > threshold ? 0 : max_value)");
-    mMode.addChoice(cv::THRESH_TOZERO, "To zero (value = value > threshold ? value : 0)");
-    mMode.addChoice(cv::THRESH_TOZERO_INV, "To zero inverted (value = value > threshold ? 0 : value)");
-    mMode.addChoice(cv::THRESH_TRUNC, "Truncate (value = value > threshold ? threshold : value)");
+    mMode.setDesc(tr("Thresholding mode"));
+    mMode.addChoice(cv::THRESH_BINARY, tr("Binary (value = value > threshold ? max_value : 0 )"));
+    mMode.addChoice(cv::THRESH_BINARY_INV, tr("Binary inverted (value = value > threshold ? 0 : max_value)"));
+    mMode.addChoice(cv::THRESH_TOZERO, tr("To zero (value = value > threshold ? value : 0)"));
+    mMode.addChoice(cv::THRESH_TOZERO_INV, tr("To zero inverted (value = value > threshold ? 0 : value)"));
+    mMode.addChoice(cv::THRESH_TRUNC, tr("Truncate (value = value > threshold ? threshold : value)"));
+    mMode.addChoice(cv::THRESH_OTSU, tr("Otsu’s Algorithm (threshold automatically determined)"));
+    mMode.addChoice(cv::THRESH_OTSU+1, tr("Otsu’s Algorithm inverted (threshold automatically determined)"));
+    mMode.addChoice(cv::THRESH_TRIANGLE, tr("Triangle algorithm (threshold automatically determined)"));
+    mMode.addChoice(cv::THRESH_TRIANGLE+1, tr("Triangle algorithm inverted (threshold automatically determined)"));
     mMode.setDefault(cv::THRESH_BINARY);
     addInputPort(mMode);
 }
 
 void ThresholdFilter::execute()
 {
     int threshold = mThreshold;
-    const GrayImage src = mIn;
-    GrayImage dest;
-    cv::threshold(src, dest, threshold, 255, mMode);
+    cv::Mat src = mIn;
+    ((Image)src).convertToGray(CV_8U);
```

Comment:
Except for Otsu’s algorithm cv::threshold should be able to handle CV_16SC1 (https://stackoverflow.com/questions/33099490/how-to-threshold-an-image-32-bit-using-otsu-s-binarization)